src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+# Understanding the problem
+
+The task consists on continuing the work done before during previous module, with the addition of playing with non-linear models to improve the performance. Moreover, we're again asked to build a MVP within a API developed by our engineering team, so it must have a fit code and a inference code which later on we will define separatedly.
+
+We will take all the decision and filters as granted from previous module, despite not being the best nor most accurate, such as:
+- Different baselines, we would use linear regression from previous module.
+- Feature selection might be more extensive.
+- Splittig sets using cumulative data instead of hard coded %'s. Even apply cross-validation addressing data leakages (ordered_before, global_popularity also if we are very nitpicky)
+
+So we will copy all previous code and start from there applying non-linear models, hence our baseline in this module would be the best model chosen in previous one (lr with ridge regularisation C=1e-5)
+
+## Import libraries
+
+
+```python
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+import logging
+from datetime import timedelta, datetime
+
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import roc_curve, auc, precision_recall_curve
+from sklearn.metrics import confusion_matrix, classification_report
+from sklearn.pipeline import Pipeline
+from sklearn.model_selection import GridSearchCV
+import joblib
+import pickle
+import os
+import json
+from typing import List, Dict, Optional
+from catboost import Pool
+from catboost import CatBoostClassifier
+import traceback
+
+```
+
+# Functions used in previous MVP
+
+
+```python
+def assess_NA(data: pd.DataFrame):
+    """
+    Returns a pd.DataFrame denoting the total number of NA
+    values and the percentage of NA values in each column.
+    """
+    # pd.Datadenoting features and the sum of their null values
+    nulls = data.isnull().sum().reset_index().rename(columns={0: "count"})
+    nulls["percent"] = nulls["count"] * 100 / len(data)
+
+    return nulls
+
+
+def read_data(file_path: str) -> pd.DataFrame:
+    """
+    Read and validate data from a CSV file.
+    Parameters: file_path (str)
+    Returns: pd.DataFrame
+    """
+    try:
+        data = pd.read_csv(file_path, index_col=False)
+
+        if data is not None and not data.empty:
```

Comment:
Small comment: I don't know if this will if statement will be ever be False, since you are already capturing: `pd.errors.EmptyDataError`